File: packages/shared-types/src/scene-graph.ts
/**
 * Canonical FloorPlanSceneGraph — the only contract consumed by
 * measurement, visualisation, and (later) compliance modules.
 * Do not treat raw model output as the application contract.
 */

export const SCENE_GRAPH_SCHEMA_VERSION = "1.0.0";

export type CoordinateSystem = "original_image_px" | "working_image_px" | "world_mm";

export type CalibrationMethod =
  | "manual_two_point"
  | "dimension_line"
  | "scale_bar"
  | "title_block_scale"
  | "manual_scale_paper"
  | "unknown";

export type EntityStatus = "predicted" | "user_confirmed" | "user_edited" | "rejected";

export type PlanEntityType =
  | "wall"
  | "door"
  | "window"
  | "room"
  | "unit_boundary"
  | "column"
  | "stair"
  | "fixture"
  | "text_label"
  | "dimension"
  | "title_block"
  | "legend"
  | "north_arrow"
  | "scale_region"
  | "notes"
  | "other"
  | "main_floorplan"
  | "drawing_border"
  | "revision_block";

export type RelationshipType =
  | "room_adjacency"
  | "room_door_access"
  | "door_to_wall"
  | "room_window_exterior"
  | "unit_to_room"
  | "room_label_assignment";

export interface Point {
  x: number;
  y: number;
}

export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export type PolylinePx = Point[];
export type PolygonPx = Point[];

export interface MaskArtifactReference {
  artifactId: string;
  mimeType?: string;
}

export interface CoordinateTransform {
  /** Maps working_image_px → original_image_px: [x', y'] = [sx*x + tx, sy*y + ty] */
  scaleX: number;
  scaleY: number;
  translateX: number;
  translateY: number;
}

export interface Evidence {
  modelId: string;
  modelVersion: string;
  sourceArtifactId: string;
  confidence: number;
  inferredAt: string;
}

export interface PlanEntity {
  id: string;
  type: PlanEntityType;
  bboxPx?: BoundingBox;
  polygonPx?: PolygonPx;
  polylinePx?: PolylinePx;
  attributes: Record<string, unknown>;
  confidence: number;
  status: EntityStatus;
  evidence: Evidence[];
  createdAt: string;
  updatedAt: string;
}

export interface EntityRelationship {
  id: string;
  type: RelationshipType;
  fromEntityId: string;
  toEntityId: string;
  confidence: number;
  attributes: Record<string, unknown>;
}

export interface Measurement {
  id: string;
  kind:
    | "room_area"
    | "room_perimeter"
    | "opening_width"
    | "wall_thickness"
    | "min_room_width"
    | "distance";
  sourceGeometryIds: string[];
  calibrationId: string;
  valuePx?: number;
  valueMm?: number;
  valueM?: number;
  valueM2?: number;
  unit: string;
  precision: number;
  confidence: number;
  formula?: string;
  estimated: boolean;
}

export interface Calibration {
  id: string;
  method: CalibrationMethod;
  mmPerPixel: number;
  confidence: number;
  sourceText?: string | null;
  sourceGeometryPx?: PolylinePx | null;
  verifiedByUser: boolean;
  active: boolean;
  createdAt: string;
}

export interface FloorPlanSceneGraph {
  schemaVersion: string;
  id: string;
  projectId: string;
  planDocumentId: string;
  pageId: string;
  analysisRunId: string;
  coordinateSystems: CoordinateSystem[];
  workingToOriginal: CoordinateTransform;
  calibration: Calibration | null;
  entities: PlanEntity[];
  relationships: EntityRelationship[];
  measurements: Measurement[];
  createdAt: string;
  updatedAt: string;
}

/** Persistent domain records (API) — UUIDs. */
export interface FpProject {
  id: string;
  name: string;
  createdAt: string;
  updatedAt: string;
}

export interface PlanDocument {
  id: string;
  projectId: string;
  filename: string;
  mimeType: string;
  byteSize: number;
  storagePath: string;
  createdAt: string;
  pages?: FpPlanPage[];
}

export interface FpPlanPage {
  id: string;
  planDocumentId: string;
  pageNumber: number;
  widthPx: number;
  heightPx: number;
  dpi: number;
  sourceFilename?: string | null;
  rasterArtifactId?: string | null;
  previewArtifactId?: string | null;
  originalImageUrl?: string | null;
  previewImageUrl?: string | null;
}

export type AnalysisRunStatus =
  | "queued"
  | "running"
  | "succeeded"
  | "partial"
  | "failed"
  | "skipped";

export interface AnalysisRun {
  id: string;
  planDocumentId: string;
  pageId: string;
  profile: string;
  status: AnalysisRunStatus;
  warning?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface Artifact {
  id: string;
  analysisRunId?: string | null;
  kind: string;
  mimeType: string;
  storagePath: string;
  byteSize: number;
  createdAt: string;
}

File: services/api/app/schemas/scene_graph.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, Field

SCENE_GRAPH_SCHEMA_VERSION = "1.0.0"

CoordinateSystem = Literal["original_image_px", "working_image_px", "world_mm"]
CalibrationMethod = Literal[
    "manual_two_point",
    "dimension_line",
    "scale_bar",
    "title_block_scale",
    "manual_scale_paper",
    "unknown",
]
EntityStatus = Literal["predicted", "user_confirmed", "user_edited", "rejected"]
PlanEntityType = Literal[
    "wall",
    "door",
    "window",
    "room",
    "unit_boundary",
    "column",
    "stair",
    "fixture",
    "text_label",
    "dimension",
    "title_block",
    "legend",
    "north_arrow",
    "scale_region",
    "notes",
    "other",
    "main_floorplan",
    "drawing_border",
    "revision_block",
]
RelationshipType = Literal[
    "room_adjacency",
    "room_door_access",
    "door_to_wall",
    "room_window_exterior",
    "unit_to_room",
    "room_label_assignment",
]


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


def new_id() -> str:
    return str(uuid4())


class Point(BaseModel):
    x: float
    y: float


class BoundingBox(BaseModel):
    x: float
    y: float
    width: float
    height: float


class MaskArtifactReference(BaseModel):
    artifact_id: str = Field(alias="artifactId")
    mime_type: str | None = Field(default=None, alias="mimeType")

    model_config = {"populate_by_name": True}


class CoordinateTransform(BaseModel):
    scale_x: float = Field(alias="scaleX")
    scale_y: float = Field(alias="scaleY")
    translate_x: float = Field(alias="translateX")
    translate_y: float = Field(alias="translateY")

    model_config = {"populate_by_name": True}


class Evidence(BaseModel):
    model_id: str = Field(alias="modelId")
    model_version: str = Field(alias="modelVersion")
    source_artifact_id: str = Field(alias="sourceArtifactId")
    confidence: float
    inferred_at: str = Field(alias="inferredAt")

    model_config = {"populate_by_name": True}


class PlanEntity(BaseModel):
    id: str
    type: PlanEntityType
    bbox_px: BoundingBox | None = Field(default=None, alias="bboxPx")
    polygon_px: list[Point] | None = Field(default=None, alias="polygonPx")
    polyline_px: list[Point] | None = Field(default=None, alias="polylinePx")
    attributes: dict[str, Any] = Field(default_factory=dict)
    confidence: float
    status: EntityStatus
    evidence: list[Evidence] = Field(default_factory=list)
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")

    model_config = {"populate_by_name": True}


class EntityRelationship(BaseModel):
    id: str
    type: RelationshipType
    from_entity_id: str = Field(alias="fromEntityId")
    to_entity_id: str = Field(alias="toEntityId")
    confidence: float
    attributes: dict[str, Any] = Field(default_factory=dict)

    model_config = {"populate_by_name": True}


class Measurement(BaseModel):
    id: str
    kind: Literal[
        "room_area",
        "room_perimeter",
        "opening_width",
        "wall_thickness",
        "min_room_width",
        "distance",
    ]
    source_geometry_ids: list[str] = Field(alias="sourceGeometryIds")
    calibration_id: str = Field(alias="calibrationId")
    value_px: float | None = Field(default=None, alias="valuePx")
    value_mm: float | None = Field(default=None, alias="valueMm")
    value_m: float | None = Field(default=None, alias="valueM")
    value_m2: float | None = Field(default=None, alias="valueM2")
    unit: str
    precision: int
    confidence: float
    formula: str | None = None
    estimated: bool = False

    model_config = {"populate_by_name": True}


class Calibration(BaseModel):
    id: str
    method: CalibrationMethod
    mm_per_pixel: float = Field(alias="mmPerPixel")
    confidence: float
    source_text: str | None = Field(default=None, alias="sourceText")
    source_geometry_px: list[Point] | None = Field(default=None, alias="sourceGeometryPx")
    verified_by_user: bool = Field(alias="verifiedByUser")
    active: bool = True
    created_at: str = Field(alias="createdAt")

    model_config = {"populate_by_name": True}


class FloorPlanSceneGraph(BaseModel):
    schema_version: str = Field(alias="schemaVersion")
    id: str
    project_id: str = Field(alias="projectId")
    plan_document_id: str = Field(alias="planDocumentId")
    page_id: str = Field(alias="pageId")
    analysis_run_id: str = Field(alias="analysisRunId")
    coordinate_systems: list[CoordinateSystem] = Field(alias="coordinateSystems")
    working_to_original: CoordinateTransform = Field(alias="workingToOriginal")
    calibration: Calibration | None = None
    entities: list[PlanEntity]
    relationships: list[EntityRelationship]
    measurements: list[Measurement]
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")

    model_config = {"populate_by_name": True}
